grpc/server.go

```diff
@@ -0,0 +1,251 @@
+				d.Logger.Debug("OverrideToReturn: %s", altConfig.OverrideToReturn)
+
+				return &emptypb.Empty{}, nil
+			}
+
+			d.Logger.Error("Endpoint %s should define either an ErrorToReturn or OverrideToReturn but does not", endptConfig.TargetEndpoint)
+		}
+	}
+
+	return handler(ctx, req)
+}
+
+// GetAlterationToPercentAffected takes a series of alterations configured for a Spec and maps them to the percentage of queries which should be affected
+func GetAlterationToPercentAffected(endpointSpecList []*pb.AlterationSpec) (map[AlterationConfiguration]PercentAffected, error) {
+	// object returned indicates, for a particular AlterationConfiguration, what percentage of queries to which it should apply
+	mapping := make(map[AlterationConfiguration]PercentAffected)
+
+	// unquantified is a holding variable used later to calculate and assign percentages to alterations which do not specify queryPercent
+	unquantifiedAlts := make([]AlterationConfiguration, 0)
+
+	pctClaimed := 0
+
+	for _, altSpec := range endpointSpecList {
+		if altSpec.ErrorToReturn == "" && altSpec.OverrideToReturn == "" {
+			return nil, status.Error(codes.InvalidArgument, "Cannot execute SendDisruption without specifying either ErrorToReturn or OverrideToReturn for all target endpoints")
+		}
+
+		if altSpec.ErrorToReturn != "" && altSpec.OverrideToReturn != "" {
+			return nil, status.Error(codes.InvalidArgument, "Cannot execute SendDisruption where ErrorToReturn or OverrideToReturn are both specified for a target endpoints")
+		}
+
+		alterationConfig := AlterationConfiguration{
+			ErrorToReturn:    altSpec.ErrorToReturn,
+			OverrideToReturn: altSpec.OverrideToReturn,
+		}
+
+		// Intuition:
+		// (1) add all endpoints where queryPercent is specified
+		// (2) track percentage of queries already claimed by an endpointConfiguration
+		// (3) track endpoints where queryPercent is missing and calculate them later
+		if altSpec.QueryPercent > 0 {
+			mapping[alterationConfig] = PercentAffected(altSpec.QueryPercent)
+			pctClaimed += int(altSpec.QueryPercent)
+		} else {
+			unquantifiedAlts = append(unquantifiedAlts, alterationConfig)
+		}
+	}
+
+	// add all endpoints where queryPercent is not specified by splitting the remaining queries equally by alteration
+	if len(unquantifiedAlts) > 0 {
+		pctPerAlt := (100 - pctClaimed) / len(unquantifiedAlts)
+
+		for i, altConfig := range unquantifiedAlts {
+			if i == len(unquantifiedAlts)-1 {
+				mapping[altConfig] = PercentAffected(100 - pctClaimed) // grab the remaining
+			} else {
+				mapping[altConfig] = PercentAffected(pctPerAlt)
+				pctClaimed += pctPerAlt
+			}
+		}
+	}
+
+	return mapping, nil
+}
+
+// GetPercentToAlteration Converts a mapping from alterationConfiguration to the percentage of requests which should return this altered response
+func GetPercentToAlteration(endpointSpecList map[AlterationConfiguration]PercentAffected, logger *zap.SugaredLogger) []AlterationConfiguration {
```

Comment:
Is there anyway to rename this function and/or the one above? `GetPercentToAlterations` vs `GetAlterationToPercentAffected` is hard for me, and hopefully users too (and it isnt just me)

Reply:
also the logger is unused here, so shouldnt be a parameter